packages/peregrine/lib/talons/CheckoutPage/PaymentInformation/useCreditCard.js

```diff
@@ -0,0 +1,440 @@
+        };
+    } else {
+        return {};
+    }
+};
+
+/**
+ * Talon to handle Credit Card payment method.
+ *
+ * @param {Boolean} props.updateButtonClicked boolean value which represents if a payment nonce request has been submitted
+ * @param {Function} props.onSuccess callback to invoke when the a payment nonce has been generated
+ * @param {Function} props.onReady callback to invoke when the braintree dropin component is ready
+ * @param {Function} props.onError callback to invoke when the braintree dropin component throws an error
+ * @param {Function} props.resetUpdateButtonClicked callback to reset the updateButtonClicked flag
+ * @param {DocumentNode} props.queries.getAllCountriesQuery query to fetch all countries data
+ * @param {DocumentNode} props.queries.getBillingAddressQuery query to fetch billing address from cache
+ * @param {DocumentNode} props.queries.getIsBillingAddressSameQuery query to fetch is billing address same checkbox value from cache
+ * @param {DocumentNode} props.queries.getPaymentNonceQuery query to fetch payment nonce saved in cache
+ * @param {DocumentNode} props.mutations.setBillingAddressMutation mutation to update billing address on the cart
+ * @param {DocumentNode} props.mutations.setCreditCardDetailsOnCartMutation mutation to update payment method and payment nonce on the cart
+ *
+ * @returns {
+ *   shouldRequestPaymentNonce: Boolean,
+ *   onPaymentError: Function,
+ *   onPaymentSuccess: Function,
+ *   onPaymentReady: Function,
+ *   isBillingAddressSame: Boolean,
+ *   countries: Object,
+ *   isDropinLoading: Boolean,
+ *   errors: Array<String>,
+ *   stepNumber: Number,
+ *   initialValues: {
+ *      firstName: String,
+ *      lastName: String,
+ *      city: String,
+ *      postalCode: String,
+ *      phoneNumber: String,
+ *      street1: String,
+ *      street2: String,
+ *      country: String,
+ *      state: String,
+ *      isBillingAddressSame: Boolean
+ *   },
+ *   shippingAddressCountry: String
+ * }
+ */
+export const useCreditCard = props => {
+    const {
+        onSuccess,
+        queries,
+        mutations,
+        onReady,
+        onError,
+        updateButtonClicked,
+        resetUpdateButtonClicked
+    } = props;
+    const {
+        getAllCountriesQuery,
+        getBillingAddressQuery,
+        getIsBillingAddressSameQuery,
+        getPaymentNonceQuery,
+        getShippingAddressQuery
+    } = queries;
+    const {
+        setBillingAddressMutation,
+        setCreditCardDetailsOnCartMutation
+    } = mutations;
+
+    /**
+     * Definitions
+     */
+
+    const [isDropinLoading, setDropinLoading] = useState(true);
+    const [shouldRequestPaymentNonce, setShouldRequestPaymentNonce] = useState(
+        false
+    );
+    /**
+     * `stepNumber` depicts the state of the process flow in credit card
+     * payment flow.
+     *
+     * `0` No call made yet
+     * `1` Billing address mutation intiated
+     * `2` Billing address mutation completed
```

Comment:
Do you ever set step 2 or 4 or 6?

For that matter could these be simplified into a single callback with some async/await usage? Using state and triggering render cycles seems excessive.

Reply:
OK I did see that we have to use multiple effects due to the nature of the braintree dropin callback. Still, steps seem to be used just for loading or not loading, so I might suggest changing them.

Reply:
It was a suggestion from @jimbo to have granular steps instead of only submit steps. Hence I broke it into granular ones.

Reply:
Ok that's fine - at least delete the unused steps.